### PR TITLE
*: remove crate_{name,type} attributes

### LINKED ISSUE #41701: Move attributes from crate source into Cargo.toml where appropriate
The compiler and std crates in-tree have a few attributes in their source code (lib.rs) that are conventionally set in the Cargo.toml file. I'd say this just wasn't cleaned up when rustbuild became the only build system, and should be cleaned up.

Possibly incomplete list of such attributes:

- `#![crate_name]`
- `#![crate_type]`

Note that some crates have different crate types in between Cargo.toml than in lib.rs (e.g., compiler crates seem to have only dylib in Cargo.toml, dylib and rlib in the source code). I'm not sure which one takes precedence, but my build directory seems to contain only dylibs and not rlibs for those crates, so I assume the Cargo.toml wins.

cc @alexcrichton -- is the above correct?

### PR DESCRIPTION
**Pull request by tamird**
Fixes #41701.

r? @arielb1

### CODE REVIEW

**Comment by tamird** [comment]
Note sure what to make of these test failures:
```
Testing libtest stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
   Compiling term v0.0.0 (file:///checkout/src/libterm)
   Compiling test v0.0.0 (file:///checkout/src/libtest)
error[E0308]: mismatched types
   --> /checkout/src/libtest/stats.rs:887:5
    |
887 | /     pub fn sum_three_items(b: &mut Bencher) {
888 | |         b.iter(|| {
889 | |             [1e20f64, 1.5f64, -1e20f64].sum();
890 | |         })
891 | |     }
    | |_____^ expected struct `__test::test::Bencher`, found struct `Bencher`
    |
    = note: expected type `fn(&mut __test::test::Bencher)`
               found type `fn(&mut Bencher) {stats::bench::sum_three_items}`

error[E0308]: mismatched types
   --> /checkout/src/libtest/stats.rs:893:5
    |
893 | /     pub fn sum_many_f64(b: &mut Bencher) {
894 | |         let nums = [-1e30f64, 1e60, 1e30, 1.0, -1e60];
895 | |         let v = (0..500).map(|i| nums[i % 5]).collect::<Vec<_>>();
896 | |
...   |
899 | |         })
900 | |     }
    | |_____^ expected struct `__test::test::Bencher`, found struct `Bencher`
    |
    = note: expected type `fn(&mut __test::test::Bencher)`
               found type `fn(&mut Bencher) {stats::bench::sum_many_f64}`

error[E0308]: mismatched types
   --> /checkout/src/libtest/stats.rs:903:5
    |
903 |     pub fn no_iter(_: &mut Bencher) {}
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `__test::test::Bencher`, found struct `Bencher`
    |
    = note: expected type `fn(&mut __test::test::Bencher)`
               found type `fn(&mut Bencher) {stats::bench::no_iter}`

error: aborting due to 3 previous errors

error: Could not compile `test`.
warning: build failed, waiting for other jobs to finish...
error: build failed


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "-j" "4" "--target" "x86_64-unknown-linux-gnu" "--release" "--locked" "--color" "always" "--manifest-path" "/checkout/src/libtest/Cargo.toml" "-p" "test:0.0.0" "-p" "term:0.0.0" "--" "--quiet"
expected success, got: exit code: 101


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
Build completed unsuccessfully in 0:23:20
Makefile:54: recipe for target 'check' failed
make: *** [check] Error 1

```

**Comment by arielb1** [comment]
The problem is this function, which redirects tests of crates with a `libtest` attribute to their own test infrastructure, instead of the builtin `libtest`.

https://github.com/rust-lang/rust/blob/6f4ab9458a7ad06c8ce630604f533c8c0c0acef4/src/libsyntax/test.rs#L611-L617

File: src/libsyntax/test.rs [hunk 1]
```diff
@@ -85,7 +85,7 @@ pub fn modify_for_testing(sess: &ParseSess,
                                            "reexport_test_harness_main");
 
     if should_test {
-        generate_test_harness(sess, resolver, reexport_test_harness_main, krate, span_diagnostic)
+        generate_test_harness(sess, crate_name, resolver, reexport_test_harness_main, krate, span_diagnostic)
```
Comment (kennytm) [inline src/libsyntax/test.rs]: This line is too long.

```
[00:03:11] tidy error: /checkout/src/libsyntax/test.rs:88: line longer than 100 chars
```

Comment (tamird) [inline src/libsyntax/test.rs]: Thank you, fixed.

**Comment by tamird** [comment]
Thanks @arielb1! This is green.

**Comment by arielb1** [comment]
These sort of changes are more of @alexcrichton's domain.

**Comment by kennytm** [comment]
`x86_64-gnu-aux` could not build `stage2-cargo`, the crate `test` not found.

```
[01:12:33] error[E0432]: unresolved import `test`
[01:12:33] 
[01:12:34] error: aborting due to previous error
[01:12:34] 
[01:12:34] error: Could not compile `cargo`.
[01:12:34] warning: build failed, waiting for other jobs to finish...
[01:12:58] error: build failed
```

**Comment by tamird** [comment]
I'm guessing that now that libtest doesn't have the name attribute, something's different in the resulting rlib (or dylib, on my mac). @alexcrichton any guidance? FWIW, it does seem like https://github.com/rust-lang/rust/pull/43994/commits/faa676964450b03648e06d186c538a20c35f84c0 is what broke this (and yet it is necessary to solve the earlier build problem).

**Comment by alexcrichton** [comment]
Looks like the bug here is that `tests/test.rs` is being compiled in Cargo which is failing to compile with the new handling. You may wish to just leave the test harness modification unchanged and leave the `#![crate_name = "test"]` annotation there. Definitely a hack, but should keep things working hopefully

**Comment by tamird** [comment]
@alexcrichton I don't follow: this failure occurs even without the mass removal of attributes (i.e. in the second commit).

**Comment by alexcrichton** [comment]
I'd recommend just backing out the modifications to libsyntax/test.rs. When doing that I'd add back `#![crate_name]` to `src/libtest/lib.rs` and that I believe should pass tests.

**Comment by tamird** [comment]
It'd be nice to understand the cause, though. I'll take another look at
this later today.

**Comment by alexcrichton** [comment]
The test harness acts differently based on its calculated value of `is_test_crate`. Today `is_test_crate` is true for libtest and `is_test_crate` is false for Cargo's unit test. With the logic you've done in this PR they both calculate to true. Presumably Cargo's unit test doens't work with `is_test_crate` being true.

**Comment by tamird** [comment]
@alexcrichton OK, done.

I don't quite understand this machinery that does specific-to-libtest stuff, but I've go ahead and documented some of this sorcery.

**Pull request merged**